lib/spack/spack/environment/environment.py

```diff
@@ -859,7 +859,13 @@ def _construct_state_from_manifest(self):
         configuration = config_dict(self.manifest)
 
         # Retrieve unification scheme for the concretizer
-        self.unify = spack.config.get("concretizer:unify", False)
+        concretizer = configuration.get("concretizer", {})
+        unify = None
+        for entry, value in concretizer.items():
+            if entry == "unify":
+                unify = value
+                break
+        self.unify = unify if unify is not None else spack.config.get("concretizer:unify", False)
```

Comment:
I'd like to double check this one. My hypothesis is that this function is called with an implicit assumption on the environment being active, in which case `spack.config` has the environment config scope added at some point of the stack.

If the issues comes from scripting instead, and we don't want to make this assumption, checking the `spack.yaml` file first, and then the configuration, might be wrong - since from a script we can still add an arbitrary amount of config layers on top of the spack.yaml single file scope

Reply:
Based on the discussion in #38240, I'd rather solve this with:
```diff
diff --git a/lib/spack/spack/environment/environment.py b/lib/spack/spack/environment/environment.py
index 82810e3b7d..6b8006ff2e 100644
--- a/lib/spack/spack/environment/environment.py
+++ b/lib/spack/spack/environment/environment.py
@@ -758,7 +758,8 @@ def __init__(self, manifest_dir: Union[str, pathlib.Path]) -> None:
 
         self.txlock = lk.Lock(self._transaction_lock_path)
 
-        self.unify = None
+        self._unify = None
+
         self.new_specs: List[Spec] = []
         self.new_installs: List[Spec] = []
         self.views: Dict[str, ViewDescriptor] = {}
@@ -782,6 +783,16 @@ def __init__(self, manifest_dir: Union[str, pathlib.Path]) -> None:
             self.manifest = EnvironmentManifestFile(manifest_dir)
             self._read()
 
+    @property
+    def unify(self):
+        if self._unify is None:
+            self._unify = spack.config.get("concretizer:unify", False)
+        return self._unify
+
+    @unify.setter
+    def unify(self, value):
+        self._unify = value
+
     def __reduce__(self):
         return _create_environment, (self.path,)
 
@@ -843,9 +854,6 @@ def _construct_state_from_manifest(self):
         else:
             self.views = {}
 
-        # Retrieve unification scheme for the concretizer
-        self.unify = spack.config.get("concretizer:unify", False)
-
         # Retrieve dev-build packages:
         self.dev_specs = copy.deepcopy(env_configuration.get("develop", {}))
         for name, entry in self.dev_specs.items():

```
The rationale is that the issue is due to Spack reading `self.unify` too early - so we can fix that by delaying when we read it. A more long term fix would be to _inject_ the globabl configuration into the environment and avoid relying on globals - but that is a major refactoring of the `Environment` class which is not required for a bugfix imho.

I wouldn't go with this solution since:
1. There are other methods in the `Environment` class that expect the global configuration to contain the manifest `SingleFileScope`, for instance `_concretize_separately`. I don't see a reason to special case this method.
2. We shouldn't be managing configuration scopes manually, when we have the `Configuration` class for that purpose

Reply:
@tldahlgren Any thoughts on :point_up: If you want, I can push a commit here in that sense.

Reply:
@alalazo Yes please do push a commit so you are credited for the work.